### PR TITLE
sliding window ops in c++

### PR DESCRIPTION
**Pull request by mywoodstock**

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/maxpool2d.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#include "ttnn/types.hpp"
+#include "ttnn/core.hpp"
+#include "ttnn/operations/core.hpp"
+#include "tt_metal/common/math.hpp"
+#include "ttnn/cpp/ttnn/operations/conv2d.hpp"
+#include "tt_eager/tt_dnn/op_library/pool/max_pool.hpp"
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/halo_op.hpp"
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/sliding_window.hpp"
+
+using namespace tt;
```
Comment (arakhmati) [inline ttnn/cpp/ttnn/operations/maxpool2d.hpp:16]: Remove `using namespace tt;` from here

File: ttnn/cpp/ttnn/operations/conv2d.hpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#include "tt_eager/tensor/tensor_utils.hpp"
+#include "tt_metal/impl/dispatch/command_queue.hpp"
+#include "tt_metal/common/math.hpp"
+#include "tt_eager/tt_dnn/op_library/pad/pad_op.hpp"
+#include "tt_eager/tt_dnn/op_library/conv/optimized_conv_op.hpp"
+#include "tt_eager/tensor/tensor.hpp"
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/sliding_window.hpp"
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/halo_op.hpp"
+
+using namespace tt;
```
Comment (arakhmati) [inline ttnn/cpp/ttnn/operations/conv2d.hpp:22]: Remove `using namespace tt;` from here

File: ttnn/cpp/ttnn/operations/maxpool2d.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace ttnn::operations {
+namespace maxpool {
```
Comment (arakhmati) [inline ttnn/cpp/ttnn/operations/maxpool2d.hpp:19]: keep the namespace same as the file name

File: ttnn/ttnn/device.py [hunk 4]
```diff
@@ -72,4 +73,14 @@ def dump_device_memory_state(device, prefix=""):
     ttl.device.DumpDeviceMemoryState(device, prefix)
 
 
+def is_wormhole_b0():
```
Comment (arakhmati) [inline ttnn/ttnn/device.py:76]: please change this function to take in `device` and figure out the arch from the device. Because what would happen if we have a heterogeneous system?

File: ttnn/ttnn/device.py [hunk 5]
```diff
@@ -72,4 +73,14 @@ def dump_device_memory_state(device, prefix=""):
...
+    ARCH_NAME = os.environ.get("ARCH_NAME", os.environ.get("TT_ARCH_NAME", "")).lower()
+    return "wormhole_b0" in ARCH_NAME
+
+
+def is_grayskull():
```
Comment (arakhmati) [inline ttnn/ttnn/device.py:81]: Same here. Make this function to take in device and figure out the arch from the device.

File: ttnn/ttnn/operations/conv2d.py [hunk 6]
```diff
@@ -4,12 +4,26 @@
 
 from typing import Tuple, Union, Dict, Optional
 import torch
```
Comment (arakhmati) [inline ttnn/ttnn/operations/conv2d.py:6]: don't import torch on top-level. Import it wherever only wherever you use it

File: ttnn/ttnn/__init__.py [hunk 7]
```diff
@@ -464,7 +465,7 @@ def manage_config(name, value):
 
 from ttnn.operations import transformer
 from ttnn.operations import kv_cache
-from ttnn.operations.conv2d import Conv2d
+from ttnn.operations.conv2d import Conv2d, conv2d, ConvConfig
```
Comment (arakhmati) [inline ttnn/ttnn/__init__.py:468]: Please rename `ConvConfig` to `Conv2dConfig`

**Review by arakhmati — CHANGES_REQUESTED** [review]

**Review by arakhmati — APPROVED** [review]

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 8]
```diff
@@ -1242,6 +1242,24 @@ void HWCommandQueue::enqueue_command(T& command, bool blocking) {
     }
 }
 
+// TODO: Currently converting page ordering from interleaved to sharded and then doing contiguous read/write
+//  Look into modifying command to do read/write of a page at a time to avoid doing copy
```
Comment (tt-aho) [inline tt_metal/impl/dispatch/command_queue.cpp:1246]: Why is this added back?

**Review by TT-BrianLiu — APPROVED** [review]

File: tt_metal/impl/program/program.hpp [hunk 9]
```diff
@@ -212,6 +216,9 @@ class Program {
 
     void add_semaphore(const CoreRangeSet & crs, uint32_t address, uint32_t init_value, CoreType core_type=CoreType::WORKER);
 
+    friend void detail::AddConfigTensor(Program &program, const Tensor& config_tensor);
+    void add_config_tensor(const Tensor& config_tensor);
+
```
Comment (tt-aho) [inline tt_metal/impl/program/program.hpp:221]: Isn't tensor an eager object and we want eager/metal to be separate?
Should we track buffers here instead?

File: tt_eager/tt_dnn/op_library/sliding_window_op_infra/utils.hpp [hunk 10]
```diff
@@ -11,6 +11,10 @@ namespace tt::tt_metal {
 
 namespace utils {
 
+inline uint32_t nearest_y(uint32_t val, uint32_t divisor) {
+    return (val + divisor - 1) / divisor * divisor;
+}
+
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/sliding_window_op_infra/utils.hpp:17]: This is a duplicate of `round_up`?

File: tt_eager/tt_dnn/op_library/sliding_window_op_infra/untilize_with_halo_config_generation_and_validation.py [hunk 11]
```diff
@@ -158,6 +158,8 @@ def decompose_conv_into_shards_and_generate_tensor_metadata(
         )
         conv_output_start_stick += conv_output_shard_height
 
+    # print("req_conv_input_shard_start_end=", req_conv_input_shard_start_end)
+
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/sliding_window_op_infra/untilize_with_halo_config_generation_and_validation.py:162]: Remove dead code

File: tt_eager/tt_dnn/op_library/sliding_window_op_infra/tt_py_untilize_with_halo.py [hunk 12]
```diff
@@ -201,6 +201,10 @@ def core_id_to_physical_coord(core_id):
                 remote_read=remote_read,
             )
 
+            # print ("padding_config", padding_config)
+            # print ("local_config", local_config)
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/sliding_window_op_infra/tt_py_untilize_with_halo.py:205]: Remove dead code

**Review by tt-aho — CHANGES_REQUESTED** [review]
I think the big thing is that this pr is trying to add an eager dependency to metal with tensors. Don't really think that is something we want

**Review by pgkeller — APPROVED** [review]
don't know anything about most of this but no concerns about the small part that overlaps w/ me be codeowner...

File: tt_metal/impl/program/program.hpp (hunk 9 above)
Comment (mywoodstock) [reply tt_metal/impl/program/program.hpp:221]: Yeah needed access to `Tensor`. You mean instead add the buffer pointer?

File: tt_eager/tt_dnn/op_library/sliding_window_op_infra/utils.hpp (hunk 10 above)
Comment (mywoodstock) [reply tt_eager/tt_dnn/op_library/sliding_window_op_infra/utils.hpp:17]: yes :) replaced

File: tt_eager/tt_dnn/op_library/sliding_window_op_infra/tt_py_untilize_with_halo.py (hunk 12 above)
Comment (mywoodstock) [reply tt_eager/tt_dnn/op_library/sliding_window_op_infra/tt_py_untilize_with_halo.py:205]: removed

File: tt_metal/impl/program/program.hpp (hunk 9 above)
Comment (mywoodstock) [reply tt_metal/impl/program/program.hpp:221]: Let me try with buffer ptr instead.

Comment (mywoodstock) [reply tt_metal/impl/program/program.hpp:221]: Previously we had a model level cache for configs, which we got rid of since we want it to be hidden from model user, so we had decided to add it to the program itself.

**Comment by mywoodstock** [comment]
> I think the big thing is that this pr is trying to add an eager dependency to metal with tensors. Don't really think that is something we want

@tt-aho I changed to use the device buffer instead of tensor and removed the tensor dependency from tt_metal. 
I think this should be good for now, if we decide to develop some infra for this on tt_eager side, we can move it then.

File: tt_metal/impl/program/program.cpp [hunk 13]
```diff
@@ -516,6 +516,11 @@ void Program::add_semaphore(const CoreRangeSet & crs, uint32_t address, uint32_t
     semaphores_.emplace_back(Semaphore( crs, address, init_value, core_type));
 }
 
+void Program::add_config_buffer(std::shared_ptr<Buffer> config_buffer) {
+    this->invalidate_compile();
```
Comment (tt-aho) [inline tt_metal/impl/program/program.cpp:520]: Why does adding the config buffer invalidate compile? Program is compiled with no knowledge of this.

**Review by tt-nshanker — APPROVED** [review]

**Review by tt-aho — COMMENTED** [review]
@abhullar-tt are you okay with this change for tracking these op config buffers in program object?

File: tt_metal/impl/program/program.hpp (hunk 9 above)
Comment (abhullar-tt) [reply tt_metal/impl/program/program.hpp]: is there some intermediate level where this info can be kept with the op?

**Review by tt-rkim — APPROVED** [review]

File: tt_metal/impl/program/program.cpp (hunk 13 above)
Comment (mywoodstock) [reply tt_metal/impl/program/program.cpp:520]: Hmm, if the op config (buffer) is different for the same program config, we want to distinguish it. But thinking about it, currently we do not have different config for same program config. So I can remove this.

**Comment by mywoodstock** [comment]
Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/9182647129
Model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9182651200

File: tt_metal/impl/program/program.hpp (hunk 9 above)
Comment (tt-aho) [reply tt_metal/impl/program/program.hpp]: This seems like it should belong with / leverage the program cache infra? Since you are associating a program with specific tensor buffers

**Comment by mywoodstock** [comment]
@tt-aho @abhullar-tt Any further objections on the program change?

**Review by abhullar-tt — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]

**Comment by mywoodstock** [comment]
model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9199932628
